File: Day-31/DuplicateInArray.py
def DuplicateUsingHashTable(arr):
    map = {}
    for i in arr:
        if i in map:
            return i
        map[i] = True

def DuplicateUsingXOR(arr):
    ans=0
    for i in arr:
        ans ^= i
    
    for i in range(0, len(arr)):
        ans ^= i
    
    return ans

print(DuplicateUsingHashTable([4, 2, 1, 3, 1]))
print(DuplicateUsingXOR([4, 2, 1, 3, 1]))

File: Day-31/DuplicateInArray.js
function duplicateInArrayHashMap(arr) {
  map = {};
  for (let i = 0; i < arr.length; i++) {
    if (map[arr[i]]) {
      return arr[i];
    }
    map[arr[i]] = true;
  }
}

console.log(duplicateInArrayHashMap([4, 2, 1, 3, 1]));

function duplicateInArrayUsingXOR(arr) {
  let ans = 0;
  for (let i = 0; i < arr.length; i++) {
    ans ^= arr[i];
  }

  for (let i = 0; i < arr.length; i++) {
    ans ^= i;
  }

  return ans;
}
console.log(duplicateInArrayHashMap([4, 2, 1, 3, 1]));
console.log(duplicateInArrayUsingXOR([4, 2, 1, 3, 1]));
